lib/chef/knife/bootstrap.rb

```diff
@@ -228,18 +228,22 @@ def knife_ssh
         ssh = Chef::Knife::Ssh.new
         ssh.ui = ui
         ssh.name_args = [ server_name, ssh_command ]
-        ssh.config[:ssh_user] = Chef::Config[:knife][:ssh_user] || config[:ssh_user]
-        ssh.config[:ssh_password] = config[:ssh_password]
-        ssh.config[:ssh_port] = Chef::Config[:knife][:ssh_port] || config[:ssh_port]
-        ssh.config[:ssh_gateway] = Chef::Config[:knife][:ssh_gateway] || config[:ssh_gateway]
-        ssh.config[:forward_agent] = Chef::Config[:knife][:forward_agent] || config[:forward_agent]
-        ssh.config[:identity_file] = Chef::Config[:knife][:identity_file] || config[:identity_file]
+        ssh.config[:ssh_user] = config_or_knife(:ssh_user)
+        ssh.config[:ssh_password] = config_or_knife(:ssh_password)
+        ssh.config[:ssh_port] = config_or_knife(:ssh_port)
+        ssh.config[:ssh_gateway] = config_or_knife(:ssh_gateway)
+        ssh.config[:forward_agent] = config_or_knife(:forward_agent)
+        ssh.config[:identity_file] = config_or_knife(:identity_file)
         ssh.config[:manual] = true
-        ssh.config[:host_key_verify] = Chef::Config[:knife][:host_key_verify] || config[:host_key_verify]
+        ssh.config[:host_key_verify] = config_or_knife(:host_key_verify)
         ssh.config[:on_error] = :raise
         ssh
       end
 
+      def config_or_knife(key)
```

Comment:
pretty certain this is all cargo cult and only config[:whatever] should be used now.  the Chef::Knife class is responsible for merging the knife config and cli variables over the Chef::Config to give a unified view.

once everything is using config[:whatever] consistently then all the Proc bullshit in the mixlib-cli options can be ripped out as well.